test/common/TestGradient.hs

```diff
@@ -337,22 +368,37 @@ dotOS :: (Numeric r, OS.Shape sh)
       => OS.Array sh r -> OS.Array sh r -> r
 dotOS arr1 arr2 = OS.toVector arr1 LA.<.> OS.toVector arr2
 
+sumOS :: (Numeric r, OS.Shape sh)
+      => OS.Array sh r -> r
+sumOS arr = sum $ OS.toList arr
+
 adoptTests :: TestTree
 adoptTests = testGroup "Tests of the port of adopt code"
   [ testCase "17.3" test_conv2d_dInp
   ]
 
 test_conv2d_dInp :: Assertion
 test_conv2d_dInp =
-  -- Below, @17.3@ is just @OS.constant 17.3@ via a @Num@ instance.
-  let arrK = 0 {-17.3-} :: OS.Array '[2, 2 {-6-}, 3, 4] Double
-      arrB = 0 {-2.28-} :: OS.Array '[5, 2, 6, 7] Double
-        -- this fails with the commented out values;
-        -- and fails to typecheck with the commented out dimension size,
-        -- because conv2d_dInp lacks the arrA parameter
-  in assertEqualUpToEpsilon 1e-7
-       (revDt (conv2d (constant arrK))  -- gradient wrt second argument
-              arrB   -- dummy, should be arrA, the point at which we take
-                     -- the gradient: https://en.wikipedia.org/wiki/Gradient
-              arrB )  -- sensitivity; should we normally test with 1?
-       (conv2d_dInp arrK arrB)  -- expected result
+  let -- Input of shape: batch x chas x height x width
+      arrA   = 1 :: OS.Array '[1, 2, 4, 4] Double
```

Comment:
A good pre-test, before I add a QuickCheck test based on your code, is to try setting all dimensions to be different, unless the names of dimension lengths are the same (or there is something like `nAcha ~ nKcha` in a function signature). We use several functions, so a couple more dimensions may have to be equated but this should be explainable. Here is where naming dimensions consistently across all functions helps a lot.